openquake/hazardlib/geo/utils.py

```diff
@@ -929,18 +929,39 @@ def geolocate(lonlats, geom_df, exclude=()):
     """
     :param lonlats: array of shape (N, 2) of (lon, lat)
     :param geom_df: DataFrame of geometries with a "code" field
+    :param exclude: List of codes to exclude from the results
     :returns: codes associated to the points
 
     NB: if the "code" field is not a primary key, i.e. there are
     different geometries with the same code, performs an "or", i.e.
     associates the code if at least one of the geometries matches
     """
     codes = numpy.array(['???'] * len(lonlats))
-    for code, df in geom_df.groupby('code'):
-        if code in exclude:
-            continue
+    filtered_geom_df = geom_df[~geom_df['code'].isin(exclude)]
+    for code, df in filtered_geom_df.groupby('code'):
         ok = numpy.zeros(len(lonlats), bool)
         for geom in df.geom:
             ok |= contains_xy(geom, lonlats)
         codes[ok] = code
     return codes
+
+
+def geolocate_geometries(geometries, geom_df, exclude=(), no_match_code='???'):
+    """
```

Comment:
You can remove the `no_match_code` if you just return the matching codes.

Reply:
Done